### PR TITLE
Improve autocompletion for zsh to be more context-aware for #356

### LINKED ISSUE #356: Add autocompletions for flags of commands when no project is required
Support for completions has just been merged in #353. However, it's not yet complete. One thing that is missing is that we don't autocomplete flags of commands *when no project is required*.
  
That is, `bloop compile foo -<TAB>` autocompletes, but `bloop configure -` or `bloop compile -` doesn't.

### PR DESCRIPTION
**Pull request by rberenguel**
Fixes #356.

### CODE REVIEW

File: etc/zsh/_bloop [hunk 1]
```diff
@@ -25,30 +27,53 @@ _boolean() {
...
+    local flags=($(bloop autocomplete --format zsh --mode flags --command $cmd))
+    flags=($flags ':approximate-4:::' )
     IFS=$saveIFS
+    _arguments $flags
+    return 0
+}
 
-    _arguments $results
+_project_or_flags() {
+    local project_bound_cmd=( clean compile run test console )
```
Comment (jvican) [inline etc/zsh/_bloop]: Would it be possible to get the list of commands that take a project from `bloop autocomplete` too? :smile:

**Comment by rberenguel** [comment]
@jvican yep, done (much better now that Scala handles that). Not very happy with how the filter-map combo look, but seems to be the most seamless way to write this. Any suggestions to make it look better?

**Comment by jvican** [comment]
Fixes #356.

**Comment by jvican** [comment]
@rberenguel Btw, just a tip: you can use "Fixes #ticket-number" to tell GitHub to close the ticket when this PR is merge.

File: frontend/src/main/scala/bloop/engine/Interpreter.scala [hunk 2]
```diff
@@ -213,9 +214,15 @@ object Interpreter {
     }
   }
 
+  private val projectBound = CommandsMessages.messages.filter {
+    case (name, CommandMessages(args, _)) => args.exists(_.name == "project")
+  }.map(_._1).mkString(" ")
```
Comment (Duhemm) [inline frontend/src/main/scala/bloop/engine/Interpreter.scala]: The filter-map combo can be replaced with `collect`:

```scala
private val projectBound = CommandsMessages.messages.filter {
  case (name, CommandMessages(args, _)) if args.exists(_.name == "projects") => name
}.mkString(" ")
```

Comment (rberenguel) [reply frontend/src/main/scala/bloop/engine/Interpreter.scala]: Thanks @Duhemm ! I knew there was to be something in the library to do this in one go

**Comment by rberenguel** [comment]
Any idea what might be up with the Windows test? `bloop.nailgun.BasicNailgunSpec` has failed twice in Windows (last two commits) but passes fine in Linux

**Comment by jvican** [comment]
> Oh thanks @jvican (btw had a facepalm with the put in a val :D ) Do you think it would make sense to actually place it in Commands.scala?
  
Yeah :smile:, I think that's the perfect place.
  
> Any idea what might be up with the Windows test? bloop.nailgun.BasicNailgunSpec has failed twice in Windows (last two commits) but passes fine in Linux
  
It's an spurious error, we just restarted the build :+1:

**Comment by rberenguel** [comment]
Moved the list of project commands to Commands. Next one to tackle will be adding (scala) tests for autocompletion :)

**Comment by jvican** [comment]
Awesome, merging :smile:

**Pull request merged**